## 1.1 Introduction

**Dataset:** Customer Churn Dataset
<br>**Source:** Kaggle
<br>**Context:** This Dataset contains the financial information of an membership firm.
<br>**Objective:** Objective is to predict risk score of the member so that organization can avert the risk fafctor.


# 1. Importing Libraries

In [5]:
import os
import warnings

import pandas as pd
import numpy as np
import seaborn as sns 

from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OrdinalEncoder


from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec

plt.style.use('ggplot')
pd.set_option('display.max_columns', None)
warnings.filterwarnings('ignore')

# 2. Reading Dataset

In [38]:
df = pd.read_parquet(r'../dataset/churn.parquet')
df.head(10).T

0                      1  \
age                                                 18                     32   
gender                                               F                      F   
security_no                                    XW0DQ7H                5K0N3X1   
region_category                                Village                   City   
membership_category                Platinum Membership     Premium Membership   
joining_date                                17-08-2017             28-08-2017   
joined_through_referral                             No                      ?   
referral_id                                   xxxxxxxx               CID21329   
preferred_offer_types            Gift Vouchers/Coupons  Gift Vouchers/Coupons   
medium_of_operation                                  ?                Desktop   
internet_option                                  Wi-Fi            Mobile_Data   
last_visit_time                               16:08:02               12:38:13   
days_since_last_login                               17                     16   
avg_time_spent                                  300.63                 306.34   
avg_transaction_value                         53005.25               12838.38   
avg_frequency_login_days                            17                     10   
points_in_wallet                                781.75                    NaN   
used_special_discount                              Yes                    Yes   
offer_application_preference                       Yes                     No   
past_complaint                                      No                    Yes   
complaint_status                        Not Applicable                 Solved   
feedback                      Products always in Stock  Quality Customer Care   
churn_risk_score                                     0                      0   

                                                  2                      3  \
age                                              44                     37   
gender                                            F                      M   
security_no                                 1F2TCL3                VJGJ33N   
region_category                                Town                   City   
membership_category                   No Membership          No Membership   
joining_date                             11-11-2016             29-10-2016   
joined_through_referral                         Yes                    Yes   
referral_id                                CID12313                CID3793   
preferred_offer_types         Gift Vouchers/Coupons  Gift Vouchers/Coupons   
medium_of_operation                         Desktop                Desktop   
internet_option                               Wi-Fi            Mobile_Data   
last_visit_time                            22:53:21               15:57:50   
days_since_last_login                            14                     11   
avg_time_spent                               516.16                  53.27   
avg_transaction_value                       21027.0               25239.56   
avg_frequency_login_days                         22                      6   
points_in_wallet                             500.69                 567.66   
used_special_discount                            No                     No   
offer_application_preference                    Yes                    Yes   
past_complaint                                  Yes                    Yes   
complaint_status                Solved in Follow-up               Unsolved   
feedback                               Poor Website           Poor Website   
churn_risk_score                                  1                      1   

                                                     4                      5  \
age                                                 31                     13   
gender                                               F                   

| __Variable__ | __Description__ |
|     :---      |       :---      |      
| __age__			 |Age of Customer in Year 			 |
| __gender__			 |Gender of Customer (M = Male; F = Female) 			 |
| __security_no__			 | Identifier 			 |
| __region_category__			 | Customer's region category:	(`Village`, `City`, `Town`, __`None`__)		 |
| __membership_category__			 | Customer's membership category: (`Platinum Membership`, `Premium Membership`, `No Membership`, `Gold Membership`, `Silver Membership`, `Basic Membership`, )		 |
| __joining_date__			 | Customer's Joining Date			 |
| __joined_through_referral__			 | whether customer joined bank through referrals 	(`No`, __`?`__, `Yes`, )		 |
| __referral_id__			 | 	refferal_id of Customer		 |
| __preferred_offer_types__			 | Customer's Preferred Offer types while purchasing: (`Gift Vouchers/Coupons`, `Credit/Debit Card Offers`, `Without Offers`, `None`, )			 |
| __medium_of_operation__			 | Device Type used by customer's while transacting	(**`?`**, `Desktop`, `Smartphone`, `Both`, )		 |
| __internet_option__			 | 	internet_option preferred by Customer 	(`Wi-Fi`, `Mobile_Data`, `Fiber_Optic`, )	 |
| __last_visit_time__			 | 	time passed since customer last visit in (hh:mm:ss)		 |
| __days_since_last_login__			 | days passed since customer last visit in days			 |
| __avg_time_spent__			 | 	Avg. Time Spent by Customer on platform		 |
| __avg_transaction_value__			 | Avg. Transaction Value of customer		 |
| __avg_frequency_login_days__			 | Avg Frequency login Days of Customer			 |
| __points_in_wallet__			 | Points in Customer Wallet			 |
| __used_special_discount__			 | 		whether customer used special discount	(`Yes`, `No`, ) |
| __offer_application_preference__			 | 	Preference of customer whether offers should be auto applied while transacting.	(`Yes`, `No`, )	 |
| __past_complaint__			 | 			Does Customer Complaint about services offered. (`No`, `Yes`, )|
| __complaint_status__			 | 		status of complaint filed by customer (`Not Applicable`, `Solved`, `Solved in Follow-up`, `Unsolved`, `No Information Available`, ) |
| __feedback__			 | 			Feedbacks of Customer |
| __churn_risk_score__			 | 			__Target Variable:__ KPI metric describes risk whether the customer will stop using services. |

# 3. Dataset Review

## 3.1 Dataset Basic Review

In [39]:
# Display a concise summary of the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36992 entries, 0 to 36991
Data columns (total 23 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           36992 non-null  int64  
 1   gender                        36992 non-null  object 
 2   security_no                   36992 non-null  object 
 3   region_category               31564 non-null  object 
 4   membership_category           36992 non-null  object 
 5   joining_date                  36992 non-null  object 
 6   joined_through_referral       36992 non-null  object 
 7   referral_id                   36992 non-null  object 
 8   preferred_offer_types         36704 non-null  object 
 9   medium_of_operation           36992 non-null  object 
 10  internet_option               36992 non-null  object 
 11  last_visit_time               36992 non-null  object 
 12  days_since_last_login         36992 non-null  int64  
 13  a

**Inferences:**
- **Count: Records:** This dataset contains **36,992** records.
- **Columns:** There are 23 Columns in the Dataset corresponding to various attributes of bank customers.
- **Data Types:**
    - **Numerical Features:** 6
        - **int64:** 3
        - **float64:** 3
    - **Categorical Features:** 17/23
        - **object:** 17
- **Missing Values:** 3 Features contains missing values. **{region_category, preferred_offer_types, points_in_wallet}**

**Note:**
- Majority of Features in Dataset is Categorical in Nature
- Some of the Categorical Features are numerical in nature needs further processing.: (`days_since_last_login`, `avg_frequency_login_days`)

## 3.2 Summary Statistics for Numeric Variables

In [42]:
#Get the summary statistics for numerical variables
df.describe().T

count          mean           std          min  \
age                    36992.0     37.118161     15.867412    10.000000   
days_since_last_login  36992.0    -41.915576    228.819900  -999.000000   
avg_time_spent         36992.0    243.472334    398.289149 -2814.109110   
avg_transaction_value  36992.0  29271.194003  19444.806226   800.460000   
points_in_wallet       33549.0    686.882199    194.063624  -760.661236   
churn_risk_score       36992.0      0.540982      0.498324     0.000000   

                              25%        50%        75%           max  
age                       23.0000     37.000     51.000     64.000000  
days_since_last_login      8.0000     12.000     16.000     26.000000  
avg_time_spent            60.1025    161.765    356.515   3235.578521  
avg_transaction_value  14177.5400  27554.485  40855.110  99914.050000  
points_in_wallet         616.1500    697.620    763.950   2069.069761  
churn_risk_score           0.0000      1.000      1.000      1.000000

__Numerical Features:__
- `age`:
    - **Avg. Age of Customer:** 37;
    - **Median Age of Customer:** 37
    - **youngest customer's Age:** 10;
    - **oldest customer's age:** 64
    - Feature seems to follow **Normal Distribution**

## 3.3 Summary Statistics for Categorical Variables

In [41]:
# Get the summary statistics for categorical variables
df.describe(include = 'object').T

count unique                    top   freq
gender                        36992      3                      F  18490
security_no                   36992  36992                YOG2LCF      1
region_category               31564      3                   Town  14128
membership_category           36992      6       Basic Membership   7724
joining_date                  36992   1096             02-06-2015     55
joined_through_referral       36992      3                     No  15839
referral_id                   36992  11359               xxxxxxxx  17846
preferred_offer_types         36704      3  Gift Vouchers/Coupons  12349
medium_of_operation           36992      4                Desktop  13913
internet_option               36992      3                  Wi-Fi  12413
last_visit_time               36992  30101                7:56:41      6
avg_frequency_login_days      36992   1654                  Error   3522
used_special_discount         36992      2                    Yes  20342
offer_application_preference  36992      2                    Yes  20440
past_complaint                36992      2                     No  18602
complaint_status              36992      5         Not Applicable  18602
feedback                      36992      9   Poor Product Quality   6350

### 1.2.3 Analyze Dataset Structure

#### 1.2.3.1 Rows & Columns

In [ ]:
print(f"Dataset contains {df.shape[0]} records and {df.shape[1]} features..!!")

#### 1.2.3.2 Removing Duplicates

In [ ]:
print(f"Dropping: Duplicate Records\t {df.duplicated().sum()}")
df = df.drop_duplicates()

#### 1.2.3.3 Features Statistics


In [ ]:
target_feature = df['churn_risk_score']

features = ['age', 'gender', 'security_no', 'region_category',
       'membership_category', 'joining_date', 'joined_through_referral',
       'referral_id', 'preferred_offer_types', 'medium_of_operation',
       'internet_option', 'last_visit_time', 'days_since_last_login',
       'avg_time_spent', 'avg_transaction_value', 'avg_frequency_login_days',
       'points_in_wallet', 'used_special_discount',
       'offer_application_preference', 'past_complaint', 'complaint_status',
       'feedback',]

# Dropping Identifiers & Textual Features
df = df.drop(columns = ['security_no', 'joining_date', 'joined_through_referral', 'referral_id', 'last_visit_time', 'feedback'])


numeric_features = df.select_dtypes(include=np.number)
categorical_features = df.select_dtypes(exclude=np.number)

In [ ]:
# Numeric Features Stats
numeric_features.describe().T

In [ ]:
# Categorical Features Stats
categorical_features.describe().T

#### 1.3.3.4 Checking NA Values

In [ ]:
# checking NA Values in numeric_cols
100*(numeric_features.isna().sum()/numeric_features.shape[0])

In [ ]:
# checking NA Values in categorical_cols
100*(categorical_features.isna().sum()/categorical_features.shape[0])

In [ ]:
# Checking NA Values in Target feature
target_feature.isna().sum()

#### 1.3.3.4.1 Ploting NA Values

In [ ]:
sns.heatmap(df.isna(), cbar = False)

**Initial Observations**<br>
- `region_category` and `points_in_wallets`has high number of null values
- `reffered_offer_types` has  less null values

- `age`: numeric col: suggests normal distribution
- `days_since_last_login`: scaler quantity: it contains -ive values
- `security_no`: identifier (must be dropped)

### 1.2.4 EDA

#### 1.2.4.1 Univariate Analysis

In [ ]:
# Numeric Features
fig = plt.figure(figsize=(12, 8))
gs = GridSpec(2, 3)

ax1 = fig.add_subplot(gs[0, 0])
sns.histplot(numeric_features['age'], kde = True, ax = ax1)

ax2 = fig.add_subplot(gs[0, 1])
sns.histplot(numeric_features['points_in_wallet'], kde = True, ax = ax2)

ax3 = fig.add_subplot(gs[0, 2])
sns.histplot(numeric_features['avg_time_spent'], kde = True, ax = ax3)

ax4 = fig.add_subplot(gs[1, 0])
sns.histplot(numeric_features['avg_transaction_value'], kde = True, ax = ax4)

ax5 = fig.add_subplot(gs[1, 1:])
sns.histplot(numeric_features['days_since_last_login'], bins = 100, kde = True, ax = ax5)


fig.set_tight_layout('tight')
plt.show()

In [ ]:
high_cardinality_cols = tuple()

fig = plt.figure(figsize=(25, 25))
gs = GridSpec(5, 4, figure=fig, wspace=0.3, hspace=0.9)

_ = []
plot_num = 0  # Track how many plots have been drawn

for col in categorical_features.columns:
    if df[col].nunique() <= 10:
        # Calculate current row and column for grid
        row_index = plot_num // 4
        col_index = plot_num % 4

        ax = fig.add_subplot(gs[row_index, col_index])
        _.append(ax)

        sns.countplot(
            x=categorical_features[col],
            ax=ax,
            order=categorical_features[col].value_counts().index
        )

        ax.bar_label(ax.containers[0])
        ax.set_xticklabels(ax.get_xticklabels(), rotation=45)
        ax.set_title(f"Count Plot: {col}")

        plot_num += 1  # Only increment if we plotted

    else:
        high_cardinality_cols += (col, )

plt.tight_layout()
plt.show()

print(f"cols with high cardinality: {high_cardinality_cols}")

- **`gender`**: suggests it has invalid values: 'unknown'
- `region_category`: `Town` > `City` > `Village`
- `membership_category`: `Basic Membership` > `No Membership` > `Gold Membership` > `Silver Membership` > `Premium Membership` > `Platinium Membership`
- **`joined through referals`** contains missing values: '?'
- `preferred_offer_types`: Gift Voucher ~ Debit Card Offers ~ Without Offers
- **`medium of operation`** has missing values: ?
- `internet_option` Wifi ~ MobileData ~ Fiber Optic
- `used_special_discount` Yes > No
- `offer_application_preference` Yes > No
- `past complaint` Yes ~ No
- `complaint_status` Not Applicable > Unresolved ~ Solved ~ Solved in Follow-up ~ No Info Available
- `feedback` Poor Product Quality ~ No reason Specified ~ Too many ads ~ Poor Webssite ~ Poor Customer Service
- High Cardinality Cols: **`joining_date`**, **`referral_id`**, **`last_visit_time`**, **`avg_frequency_login_days`**

#### 1.2.4.2 Cleaning Dataset

##### 1.2.4.2.1 Cleaning Feature: avg_frequency_login_days

In [ ]:
def check_numeric_val(val):
    '''This Function will return Numeric Value if string is numeric in nature else return np.nan'''
    try: 
        return float(val)
    except Exception as e:
        return np.nan
        
# Preprocessing last_visit_time
# Typecasting to int
df['avg_frequency_login_days'] = df['avg_frequency_login_days'].map(check_numeric_val)


In [ ]:
fig = plt.figure(figsize=(12, 8))
gs = GridSpec(2, 1)

ax1 = fig.add_subplot(gs[0, 0])
ax1.set_title('Distribution: Avg_login frequency')
sns.histplot(x = df['avg_frequency_login_days'], ax = ax1, kde = True)

ax2 = fig.add_subplot(gs[1, 0])
ax2.set_title('Outliers in Avg login frequency')
sns.boxplot(data = df, 
            x = 'avg_frequency_login_days', ax = ax2)

plt.tight_layout()
plt.show()

In [ ]:
# Updating Avg_frequency_login_days where -ive values
df['avg_frequency_login_days'] = df['avg_frequency_login_days'].map(lambda x: np.nan if x<0 else x)

fig = plt.figure(figsize=(12, 8))
gs = GridSpec(2, 1)

ax1 = fig.add_subplot(gs[0, 0])
ax1.set_title('Distribution: Avg_login frequency')
sns.histplot(x = df['avg_frequency_login_days'], ax = ax1, kde = True)

ax2 = fig.add_subplot(gs[1, 0])
ax2.set_title('Outliers in Avg login frequency')
sns.boxplot(data = df, 
            x = 'avg_frequency_login_days', ax = ax2)

plt.tight_layout()
plt.show()

#### Preprocessing Column: 'Gender'

In [ ]:
print(df['gender'].value_counts())

# Dropping Rows gender = 'Unknown'
print(f'\n\t- Dropping Rows where Gender = "Unknown" (59 Records)..!\n')
df = df.drop(
    index = df.loc[df['gender'] == 'Unknown'].index
).reset_index(drop = True)

le = LabelEncoder()
df['gender_encoded'] = le.fit_transform(df['gender'])

print(f"\t- Feature Encoded: ['F' : 0, 'M' : 1]")

#### Preprocessing Column: Region_Category

In [ ]:
print(df['region_category'].value_counts())

sns.countplot(data = df, x = 'region_category', hue = 'churn_risk_score')
plt.show()
print(f"\t- Above graph shows clear order between region")

region_encoded = pd.get_dummies(df['region_category'], prefix='region_category_encoded_', drop_first=True, dtype=int)

df = pd.concat((df, region_encoded), axis = 1)


#### Preprocessing membership_category

In [ ]:
print(f"NULL Values: membership_category: {df['membership_category'].isna().sum()}")
print(f"\n{df['membership_category'].value_counts()}\n")
plt.figure(figsize=(5, 5))
ax = sns.barplot(data = df, y = 'churn_risk_score', x = 'membership_category', estimator='mean')
ax.bar_label(container=ax.containers[0])

plt.title('Barplot of Membership_category')
plt.xticks(rotation=45 )
plt.tight_layout()
plt.show()

In [ ]:
encoded_membership_category = pd.get_dummies(df['membership_category'], prefix='membership_category_encoded_', drop_first=1, dtype=int)
df = pd.concat((df, encoded_membership_category), axis = 1)

#### Preprocessing preferred_offer_types

In [ ]:
print(f"NULL Values: preferred_offer_types: {df['preferred_offer_types'].isna().sum()}")
print(f"\n{df['preferred_offer_types'].value_counts()}\n")
plt.figure(figsize=(5, 5))
ax = sns.barplot(data = df, y = 'churn_risk_score', x = 'preferred_offer_types', estimator='mean')
ax.bar_label(container=ax.containers[0])

plt.title('Barplot of preferred_offer_types')
plt.xticks(rotation=45 )
plt.tight_layout()
plt.show()

In [ ]:
# One-hot encoding
prefered_offered_types = pd.get_dummies(df['preferred_offer_types'], prefix='offer_type', drop_first=True, dtype = int)
df = pd.concat((df, prefered_offered_types), axis = 1)

#### Preprocessing medium_of_operation

In [ ]:
print(f"NULL Values: medium_of_operation: {df['medium_of_operation'].isna().sum()}")
print(f"\n{df['medium_of_operation'].value_counts()}\n")
plt.figure(figsize=(5, 5))
ax = sns.barplot(data = df, y = 'churn_risk_score', x = 'medium_of_operation', estimator='mean')
ax.bar_label(container=ax.containers[0])

plt.title('Barplot of medium_of_operation')
plt.xticks(rotation=45 )
plt.tight_layout()
plt.show()

print(f"\t- Dropping NA Values: medium of operation")
df = df.dropna(subset = 'medium_of_operation', ignore_index=True)


In [ ]:
medium_of_operation = pd.get_dummies(df['medium_of_operation'], dtype = int, prefix='medium_of_operation_encoded_', drop_first=True)
df = pd.concat((df, medium_of_operation), axis = 1)


#### Preprocessing internet_operation

In [ ]:
print(f"NULL Values: internet_option: {df['internet_option'].isna().sum()}")
print(f"\n{df['internet_option'].value_counts()}\n")
plt.figure(figsize=(5, 5))
ax = sns.barplot(data = df, y = 'churn_risk_score', x = 'internet_option', estimator='mean')
ax.bar_label(container=ax.containers[0])

plt.title('Barplot of internet_option')
plt.xticks(rotation=45 )
plt.tight_layout()
plt.show()


internet_option = pd.get_dummies(df['internet_option'], dtype = int, prefix='internet_option_encoded_', drop_first=True)
df = pd.concat((df, internet_option), axis = 1)


In [ ]:

ax = sns.displot(x = df['days_since_last_login'], kde = True)
ax.set_titles('Feature: days_since_last_login before Processing')

print(f"\t- Feature: days_since_last_login: Scaler Unit taking mod..!!")
df['days_since_last_login_processed'] = df['days_since_last_login'].abs()


ax2 = sns.displot(x = df['days_since_last_login_processed'], kde = True)
ax2.set_titles('Feature: days_since_last_login after Processing')
plt.show()


In [ ]:

ax = sns.displot(x = df['avg_time_spent'], kde = True, hue = df['churn_risk_score'])
ax.set_titles('Feature: avg_time_spent before Processing')

print(f"\t- Feature: avg_time_spent: Scaler Unit taking mod..!!")
df['avg_time_spent_processed'] = df['avg_time_spent'].abs()


ax2 = sns.displot(x = df['avg_time_spent_processed'], kde = True, hue = df['churn_risk_score'])
ax2.set_titles('Feature: avg_time_spent after Processing')
plt.show()


In [ ]:
sns.displot(x = df['avg_transaction_value'], kde = True, hue = df['churn_risk_score'], ax = ax1)
plt.title('Feature: avg_transaction_value before Processing')
plt.show()

In [ ]:
sns.boxplot(y = df['avg_frequency_login_days'], hue = df['churn_risk_score'])
